### PR TITLE
sql: implement || operator for json

### PR DESCRIPTION
**Pull request by justinj**
The Postgres docs for this operator can be seen here:
https://www.postgresql.org/docs/9.5/static/functions-json.html

Release note: Add || operator for JSONB.

### CODE REVIEW

**Comment by m-schneider** [comment]
Look good just one nit. :lgtm: 

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/util/json/json.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20689#-L0GmW2QI0b3s2_fDhWM:-L0GmW2QI0b3s2_fDhWN:b-y29q89) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cdf35a6371d780abe78f3714039160c5598f157/pkg/util/json/json.go#L825)):*
> ```Go
> 		right := decoded.(jsonArray)
> 		result := make([]interface{}, len(j)+len(right))
> 		for i := range j {
> ```

nit: This makes a lot of sense, but for readability can you rename j to left please and name len(j).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20689#-:-L0Go2ThMuUnZakstAHO:bpl7afk)*

**Comment by justinj** [comment]
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/util/json/json.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20689#-L0GmW2QI0b3s2_fDhWM:-L0KvZNT4ugmtWNCs3dm:b64z872) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cdf35a6371d780abe78f3714039160c5598f157/pkg/util/json/json.go#L825)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

nit: This makes a lot of sense, but for readability can you rename j to left please and name len(j).
</blockquote></details>

Done. Good call! I think link will complain if I name the receiver `left` but I will alias it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20689)*

**Comment by justinj** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/util/json/json.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20689#-L0GmW2QI0b3s2_fDhWM:-L0KxAtifVMGM04l-mdy:b-8nj3co) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cdf35a6371d780abe78f3714039160c5598f157/pkg/util/json/json.go#L825)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Done. Good call! I think link will complain if I name the receiver `left` but I will alias it.
</blockquote></details>

*lint

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20689)*

**Review by jordanlewis — COMMENTED** [review]
Nice, this is cool!

File: pkg/sql/logictest/testdata/logic_test/json_builtins [hunk 1]
```diff
@@ -404,3 +404,13 @@ query T
...
+
+query T
+SELECT '[1,2,3]'::JSON || '[4,5,6]'::JSON
+----
+[1,2,3,4,5,6]
+
+query T
+SELECT '{"a": 1, "b": 2}'::JSON || '{"b": 3, "c": 4}'::JSON
+----
+{"a":1,"b":3,"c":4}
```
Comment (jordanlewis) [inline pkg/sql/logictest/testdata/logic_test/json_builtins]: What about error cases? We should test that invalid concats produce the correct error.
Also, it's worth noting that in Postgres you don't need to cast the second argument to JSON. Do we support that?

File: pkg/util/json/json.go [hunk 2]
```diff
@@ -769,6 +784,98 @@ func (jsonFalse) RemoveIndex(int) (JSON, error)  { return nil, errCannotDeleteFr
 func (jsonString) RemoveIndex(int) (JSON, error) { return nil, errCannotDeleteFromScalar }
 func (jsonNumber) RemoveIndex(int) (JSON, error) { return nil, errCannotDeleteFromScalar }
 
+var errInvalidConcat = pgerror.NewError(pgerror.CodeInvalidParameterValueError, "invalid concatenation of jsonb objects")
```
Comment (jordanlewis) [inline pkg/util/json/json.go]: Need to test that this gets returned at the correct moments.

File: pkg/util/json/json_test.go [hunk 3]
```diff
@@ -881,6 +881,95 @@ func TestEncodeDecodeJSONInvertedIndex(t *testing.T) {
...
+			{`null`, `[false, null]`},
+			{`[]`, `[false]`},
+			{`[1]`, `[false, 1]`},
+			{`{}`, expectError},
+		},
+		`123`: {
+			{`null`, `[123, null]`},
+			{`[]`, `[123]`},
+			{`[1]`, `[123, 1]`},
+			{`{}`, expectError},
```
Comment (jordanlewis) [inline pkg/util/json/json_test.go]: What about concatting a scalar?
```
jordan=# select '1'::jsonb || '1';
?column?

File: pkg/util/json/json_test.go [hunk 4]
```diff
@@ -881,6 +881,95 @@ func TestEncodeDecodeJSONInvertedIndex(t *testing.T) {
...
+		},
```
Comment (jordanlewis) [inline pkg/util/json/json_test.go]: [1, 1]
(1 row)
```

File: pkg/util/json/json_test.go [hunk 5]
```diff
@@ -881,6 +881,95 @@ func TestEncodeDecodeJSONInvertedIndex(t *testing.T) {
...
+			{`[]`, `["hello"]`},
+			{`[1]`, `["hello", 1]`},
+			{`{}`, expectError},
+		},
+		`[1]`: {
+			{`null`, `[1, null]`},
+			{`[2]`, `[1, 2]`},
+			{`[2, 3]`, `[1, 2, 3]`},
+			{`{"a": 1}`, `[1, {"a": 1}]`},
+		},
```
Comment (jordanlewis) [inline pkg/util/json/json_test.go]: Here too - what about concatting a scalar to this array?

File: pkg/sql/logictest/testdata/logic_test/json_builtins (hunk 1 above)
Comment (justinj) [reply pkg/sql/logictest/testdata/logic_test/json_builtins]: Added an error case, and you're correct, it's not necessary!

File: pkg/util/json/json_test.go (hunk 3 above)
Comment (justinj) [reply pkg/util/json/json_test.go]: `null` is a scalar! But I've added some other scalars as well.

**Review by jordanlewis — APPROVED** [review]
LGTM

**Comment by justinj** [comment]
TFTR!

**Pull request merged**